lib/chai/core/assertions.js

```diff
@@ -1412,26 +1536,64 @@ module.exports = function (chai, _) {
   Assertion.addMethod('within', function (start, finish, msg) {
     if (msg) flag(this, 'message', msg);
     var obj = flag(this, 'object')
-      , range  = start + '..' + finish
       , doLength = flag(this, 'doLength')
       , flagMsg = flag(this, 'message')
-      , ssfi = flag(this, 'ssfi');
+      , msgPrefix = (flagMsg) ? flagMsg + ': ' : ''
+      , ssfi = flag(this, 'ssfi')
+      , objType = _.type(obj).toLowerCase()
+      , startType = _.type(start).toLowerCase()
+      , finishType = _.type(finish).toLowerCase()
+      , range = (startType === 'date' && finishType === 'date')
+          ? start.toUTCString() + '..' + finish.toUTCString()
+          : start + '..' + finish;
 
     if (doLength) {
       new Assertion(obj, flagMsg, ssfi, true).to.have.property('length');
     } else {
-      new Assertion(obj, flagMsg, ssfi, true).is.a('number');
+      if (!(objType === 'number' || objType === 'date')) {
+        obj = (objType === 'string') 
+          ? '\'' + obj + '\''
+          : obj;
+        throw new AssertionError(
+          msgPrefix + 'expected ' + obj + ' to be a number or a date',
+          undefined,
+          ssfi
+        );
+      }
     }
 
-    if (typeof start !== 'number' || typeof finish !== 'number') {
-      flagMsg = flagMsg ? flagMsg + ': ' : '';
+    var startFailedTypeCheck = !(startType === 'number' || startType === 'date');
+    var finishFailedTypeCheck = !(finishType === 'number' || finishType === 'date');
+
+    if (startFailedTypeCheck || finishFailedTypeCheck) {
+      var shoudBeType = (objType === 'number' || objType === 'date' )
+        ? objType + 's'
+        : 'numbers or dates';
       throw new AssertionError(
-        flagMsg + 'the arguments to within must be numbers',
+        msgPrefix + 'the arguments to within must be ' + shoudBeType,
         undefined,
         ssfi
       );
     }
 
+    if (doLength && (startType === 'date' || finishType === 'date')) {
+      throw new AssertionError(
+        msgPrefix + 'the arguments to within must be numbers',
+        undefined,
+        ssfi
+      );
+    }
+
+    if (!doLength && (objType !== startType || objType !== finishType)) {
+      throw new AssertionError(
+        msgPrefix + 'type mismatch, expected to within values to be ' + objType + 's',
+        undefined,
+        ssfi
+      );
+    }
+
+    // Should it check if before <= after?
```

Comment:
I noticed that within method does not check weather the two given params make a valid range. Should it? Or it's assumed that it's gonna always throw and the user would see it from the generic error message?